app/src/main/java/org/mozilla/fenix/HomeActivity.kt

```diff
@@ -723,5 +718,23 @@ open class HomeActivity : LocaleAwareAppCompatActivity(), NavHostActivity {
         // PWA must have been used within last 30 days to be considered "recently used" for the
         // telemetry purposes.
         const val PWA_RECENTLY_USED_THRESHOLD = DateUtils.DAY_IN_MILLIS * 30L
+
+        /**
+         * set a delay until we locate visual completeness queue.
+         * static function to avoid memory leak in case we are trying to access visualCompletenessQueue
+         * or other activity members when activity is already destroy.
+         */
+        private fun setupRootContainerPreDrawDelay(
```

Comment:
Not that it needs to change but why did you decide to put this in `HomeActivity` rather than a new file or a file related to the visual completeness queue?

If it makes a difference, I think the `RunWhenReadyQueue` is named in a way that makes it prematurely generic (e.g. what else do we use it for?) so I would support renaming it to the `VisualCompletenessQueue` if that helps (and it'd be easier to find in code search too).

Keeping things simple makes them easier to change and prematurely making things generic, meaning you have to support use cases that you're not actively using them for yet, makes things harder to change. Most times the prematurely generic thing is never repurposed or the requirements for the generic object change so it needs to be refactored anyway.

---

Also note: I don't usually like putting static functions in `companion` blocks of a class that has member functions. It's a usually code smell that some bit of functionality unrelated to the current class is getting piggy-backed on the current class and can be better named somewhere else (e.g. in this case, it's doing visualCompleteness related work – it's attaching the visual completness queue ready up to respond to the UI state).

This is a problem because:
- It can make the functionality harder to find
- It increases the size of the class, making it harder to skim to identify what the class is about
- It often violates the single-responsibility principle for the class
- It makes the `companion object` (unnecessarily) longer

I think what you wrote is probably the only time where I I feel less strongly about it – it could live here but it needs isolation from the capturing.

But in general, Activities are bloated classes and it's good to get as much code out of them, and into better encapsulations, as possible. e.g. HomeActivity is 740 lines, many of which are low-level things, when its primary responsibility is passing high-level lifecycle events (onResume, onStop, etc.). It seems better to include it in a 70 line low-level visual completeness queue file than here. (Note: fenix doesn't often do this but I think they should)

Reply:
> Keeping things simple makes them easier to change and prematurely making things generic, meaning you have to support use cases that you're not actively using them for yet, makes things harder to change. Most times the prematurely generic thing is never repurposed or the requirements for the generic object change so it needs to be refactored anyway.

I looked through the history and it looks like this was refactored so it could be re-used (i.e. it's not prematurely generic). I wonder what the best way to merge our needs with the class defined in a-c is.

Reply:
@mcomella  Would adding a class inside  `PerformanceComponent` file to deal with this make sense? I was a bit hesitant to just randomly create a file in the components section. I was thinking maybe the classes can co-exist in the same place. If not, I can add another file.